src/test/groovy/unit/org/pih/warehouse/api/ProductApiControllerSpec.groovy

```diff
@@ -0,0 +1,121 @@
+package unit.org.pih.warehouse.api
+
+import grails.orm.PagedResultList
+import grails.testing.gorm.DataTest
+import grails.testing.web.controllers.ControllerUnitTest
+import java.util.stream.IntStream
+import org.grails.plugins.testing.GrailsMockHttpServletResponse
+import org.grails.web.json.JSONElement
+import org.grails.web.json.JSONObject
+import org.hibernate.Criteria
+import org.json.JSONArray
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Unroll
+
+import org.pih.warehouse.api.ProductApiController
+import org.pih.warehouse.core.Location
+import org.pih.warehouse.forecasting.ForecastingService
+import org.pih.warehouse.product.Product
+import org.pih.warehouse.product.ProductService
+
+@Unroll
+class ProductApiControllerSpec extends Specification implements DataTest, ControllerUnitTest<ProductApiController> {
+
+	@Shared private def forecastingServiceStub
+	@Shared private def productServiceStub
+
+    void setupSpec() {
+		mockDomains(Product, Location)
+    }
+
+	void setup() {
+		productServiceStub = Stub(ProductService)
+		forecastingServiceStub = Stub(ForecastingService)
+
+		controller.forecastingService = forecastingServiceStub
+		controller.productService = productServiceStub
+	}
+
+	void 'getDemand should return correctly'() {
+		given: 'the following db data'
+		Product product = new Product().save(validate: false)
+		Location location = new Location().save(validate: false)
+
+		and: 'the following params'
+		String productIdToLookFor = shouldFindProduct ? product.id : '-1'
+		String locationIdToLookFor = shouldFindLocation ? location.id : '-1'
+
+		params.id = productIdToLookFor
+		session.warehouse = [id: locationIdToLookFor]
+
+		and: 'the following mocks'
+		forecastingServiceStub.getDemand(_, _, _) >> [:]
+
+		when:
+		controller.demand()
+
+		then:
+		JSONObject json = getJsonObjectResponse(controller.response)
+		if (shouldFindLocation) json.data.location != null
+		else json.data.location == null
+
+		if (shouldFindProduct) json.data.product != null
+		else json.data.product == null
+
+		json.data.demand == [:]
```

Comment:
wasn't it supposed to initialize the `json.data.demand` with an empty Map? This line just checks if `json.data.demand` is equal to the empty map.

Reply:
It's supposed to be an == because I'm testing that it returned the data from the mocking that I did in `forecastingServiceStub.getDemand(_, _, _) >> [:]`. I suppose I could mock something more interesting there though so it's more obvious that it's actually returning something.

Reply:
@EWaterman I haven't written many tests, so I didn't notice it is a shortcut of the `assert`. If it is the case, then I'm fine with that.